docs/proposals/20250513-propogate-taints.md

```diff
@@ -0,0 +1,467 @@
+This is especially useful in scale-from-zero scenarios, where that autoscaling technology can reference taints on a collection to make decisions about the cloud resources available.
+
+#### Story 3
+
+As a user, I would like to update taint metadata on my collection resources without forcing a complete replacement of an owned resource, such as a Machine or Node.
+
+### Requirements (Optional)
+
+#### Functional Requirements
+
+Functional requirements are the properties that this design should include.
+
+##### FR1
+
+Users should be able to define Taints on collection resources and have the taints propagate to the owned resources.
+This would start at a `ClusterClass` or `Cluster` level, and ultimately be written to a `Node`.
+
+##### FR2
+
+Users should be able to remove Taints managed by Cluster API without removing taints that Cluster API does not manage.
+
+### Implementation Details/Notes/Constraints
+
+#### The `.spec.taints` field on a `Node`
+
+Cluster API already supports propagating labels and annotations downward in its resource heirarchy.
+This support is implemented such that when these fields are updated, the underlying compute resources are _not_ replaced.
+
+Taints present a challenge to this, because they are defined as an "atomic" field by Kubernetes.[^1]
+This means that when updating Taints on a `Node`, _all_ `Taint`s are replaced; it is not possible to add and replace individual elements like labels and annotations.
+As a concrete example, if a Node has 3 taints and some client submits a patch request with only one, the end result is one taint on the Node.
+It also means that Server-Side Apply ownership rules could not be applied to individual taints, which could present conflicts between controllers or users trying to modify `Taints` on the resultant `Node`.
+
+For Cluster API to support propagating `Taint`s, it will need to:
+
+- implement its own mechanism for tracking what `Taint`s it owns. This will be similar to the implementation for annotations and labels. See [Changes to the Machine and MachinePool controllers](#changes-to-the-machine-and-machinepool-controllers) for more details.
+- ensure there are no conflicts with other actors by always setting the `metadata.resourceVersion` on API calls changing the taints on a Node.
+
+#### Propagation of taints
+
+A taint for a Node may be defined for different use-cases:
+
+- Taints supposed to stay on the Node to ensure only certain workload runs on such a `Node` aka. `Always`:
+  - These taints are supposed to be set on the `Node` object as long as it is defined on its parent core CAPI object.
+  - Example: Nodes where only GPU related workload should run
+  - CAPI controllers should ensure via reconciliation that such a taint gets added again in case a third party removes it.
+- Taint's supposed to get added once to a Node aka. `Initialize`
+  - This taints are supposed to be set **once** on the `Node` object.
```

Comment:
TBD when we implement, but we have to figure it out how the code can determine "once" in a re-entrant way.

Just to mention it, we can also consider to delegate adding taint on create to bootstrap providers, or a mixed solution